Source/WebP/WebPDataDecoder.m

```diff
@@ -0,0 +1,174 @@
+//  WebPDataDecoder.m
+//  Nuke-WebP-Plugin iOS
+//
+//  Created by nagisa-kosuge on 2018/04/30.
+//  Copyright © 2018年 RyoKosuge. All rights reserved.
+//
+
+#import "WebPDataDecoder.h"
+#import "webp/decode.h"
+
+void free_image_data(void *info, const void *data, size_t size) {
+    if (info != NULL) {
+        WebPFreeDecBuffer(&((WebPDecoderConfig *) info)->output);
+        free(info);
+    }
+    
+    WebPFree((void *)data);
+}
+
+CGColorSpaceRef NWPCGColorSpaceGetDeviceRGB(void) {
+    static CGColorSpaceRef colorSpace;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        colorSpace = CGColorSpaceCreateDeviceRGB();
+    });
+    return colorSpace;
+}
+
+
+@implementation WebPDataDecoder {
+    WebPIDecoder *_idec;
+}
+
+- (void)dealloc {
+    if (_idec) {
+        WebPIDelete(_idec);
+        _idec = NULL;
+    }
+}
+
+- (UIImage *)incrementallyDecodeData:(NSData *)data isFinal:(BOOL)isFinal {
+
+    if (!_idec) {
+        _idec = WebPINewRGB(MODE_rgbA, NULL, 0, 0);
+        if (!_idec) {
+            return nil;
+        }
+    }
+
+    Image *image;
+
+    VP8StatusCode status = WebPIUpdate(_idec, [data bytes], [data length]);
+    if (status != VP8_STATUS_OK && status != VP8_STATUS_SUSPENDED) {
+        return nil;
+    }
+
+    int width, height, last_y, stride = 0;
+    uint8_t *rgba = WebPIDecGetRGB(_idec, &last_y, &width, &height, &stride);
+
+    if (0 < width + height && 0 < last_y && last_y <= height) {
+        size_t rgbaSize = last_y * stride;
+        CGDataProviderRef provider = CGDataProviderCreateWithData(NULL, rgba, rgbaSize, NULL);
+        CGColorSpaceRef colorSpaceRef = NWPCGColorSpaceGetDeviceRGB();
+        CGBitmapInfo bitmapInfo = kCGBitmapByteOrder32Big | kCGImageAlphaPremultipliedLast;
+        size_t components = 4;
+        CGColorRenderingIntent renderingIntent = kCGRenderingIntentDefault;
+
+        CGImageRef imageRef = CGImageCreate(width, last_y, 8, components * 8, components * width, colorSpaceRef, bitmapInfo, provider, NULL, NO, renderingIntent);
+
+        CGDataProviderRelease(provider);
+
+        if (!imageRef) {
+            return nil;
+        }
+
+        CGColorSpaceRef canvasColorSpaceRef = NWPCGColorSpaceGetDeviceRGB();
+        CGContextRef canvas = CGBitmapContextCreate(NULL, width, height, 8, 0, canvasColorSpaceRef, bitmapInfo);
+        if (!canvas) {
+            CGImageRelease(imageRef);
+            return nil;
+        }
+
+        CGContextDrawImage(canvas, CGRectMake(0, height - last_y, width, last_y), imageRef);
+        CGImageRef newImageRef = CGBitmapContextCreateImage(canvas);
+
+        CGImageRelease(imageRef);
+        if (!newImageRef) {
+            CGContextRelease(canvas);
+            return nil;
+        }
+
+#if WEBP_IMAGE_MAC
+        image = [[NSImage alloc] initWithCGImage: imageRef size: CGSizeZero];
+#else
+        image = [UIImage imageWithCGImage:imageRef];
+#endif
+        CGImageRelease(newImageRef);
+    }
+
+    if (isFinal) {
```

Comment:
This step isn't really necessary, ImagePipeline immediately deallocates a decoder object the moment the final image for a request is produced. So right after the final image is produced `WebPDataDecoder.deallocate` is going to be called which a good place to cleanup used resources.

Reply:
@kean understood!
thank you!